### PR TITLE
Fix gan to adapt callback API changes

### PR DESCRIPTION
**Pull request by PPPW**
Fix GAN to adapt the changes in commit "Change in the Callback API". Works fine if validation set is not empty. 

If the validation set is empty, although training works fine, the `gen_loss` and `disc_loss` still won't be added to `last_metrics` and won't show in the progress bar. The `gen_loss` and `disc_loss` are added in `on_epoch_end` of `GANTrainer`, if `last_metrics` is not `None`, (with my changes here) the two losses will be added to it and recorded (the `Recorder` is called after `GANTrainer`); But when the validation set is empty, `self.state_dict['last_metrics']` is set to None in `on_epoch_end` of `CallbackHandler`, then I'm not sure how to add the GAN metrics to it. 
Perhapse we shouldn't set `self.state_dict['last_metrics']` to `None` (but set to an empty list instead), so all callbacks can still share and modify it?

### CODE REVIEW

**Comment by sgugger** [comment]
You should use `add_metrics`that will directly do that check for you and return an updated list.
Thanks in any case, I had forgotten those and this PR is really needed!

**Comment by PPPW** [comment]
Hi @sgugger, right, I saw `add_metrics` in the `Recorder` was removed which causes this error. Do you mean we can now use the `add_metrics` in `torch_core`? It checks for `None`, yes. But if `last_metrics` is `None` in `on_epoch_end` of `GANTrainer` (only happens when valid set is empty), my solution here does't work, and I don't see how to make it work without further changes. May I missed something? 

If `last_metrics` is an empty list, I can modify it in `GANTrainer` and the `Recorder` will use the updated list, it's like pass by reference; but if it's `None` and I reset it to empty list, in the `Recorder` it's still `None`, so this trick won't work anymore...

**Comment by sgugger** [comment]
The `GANTrainer` callback happens before the `Recorder` (see the [order](https://github.com/fastai/fastai/blob/b6986e1dbf8b2a52e19e6c12e5f90131cdb90825/fastai/vision/gan.py#L74)) so if it returns an updated `last_metrics` list, this is what `Recorder` will receive.
I'm not sure of the problem you think there is here?

**Comment by PPPW** [comment]
Sorry for the confusion. In `_call_and_update` of the `CallbackHandler`, the `state_dict` is passed into each callback, in the `order` as you mentioned. But the updates of `last_metrics` won't be got by later callbacks if it's `None`. In the following example: 

Untitled

My understanding may be wrong, there may be better ways to handle this, but this is what the concern is, hope I made it clearer, as there're quite a few changes going on... 

Anyway, the problem I'm concerned about is, if I train a GAN with empty validation set, the `gen_loss` and `disc_loss` are not showing in the console, because they are added in the `GANTrainer` but it's still `None` in the `Recorder`, for the reason explained above. I wonder could you replicate this issue?

**Comment by sgugger** [comment]
It'll be easier to show you. Let me merge this and I'll update the line.

**Comment by sgugger** [comment]
Done in [this commit](https://github.com/fastai/fastai/commit/3305fd33dd5e7cc4e07806929396e776e25f934d). By returning the result of `add_metrics` (which checks for None), we make sure the metrics are updated and sent to the recorder.

**Comment by PPPW** [comment]
😅... I see, that's how it works... Things get updated in the `_call_and_update` of the dict `new`. I'm still thinking in the old mechanism. Thank you for the explanation! 

Then I think my line in the `on_epoch_end` of `GANTrainer` should be removed (first line), it's redundant and will add metrics twice

**Comment by sgugger** [comment]
Oh yes, will do!

**Comment by PPPW** [comment]
One minor issue again: it seems “add_metrics” only handles None “mets”, not None “last_metrics”... When the validation set is empty we got None “last_metrics”, so we need to check for that somewhere as well...
